File: frontend/lib/types/audit.ts
/**
 * Audit Log Types
 * Type definitions for activity tracking and audit trail
 */

export enum AuditAction {
  CREATE = 'CREATE',
  UPDATE = 'UPDATE',
  DELETE = 'DELETE',
  LOGIN = 'LOGIN',
  LOGOUT = 'LOGOUT',
  ACCESS = 'ACCESS',
  EXPORT = 'EXPORT',
  IMPORT = 'IMPORT',
  GENERATE = 'GENERATE',
  SCAN = 'SCAN',
  CHECKOUT = 'CHECKOUT',
  CHECKIN = 'CHECKIN',
  RESERVE = 'RESERVE',
  RENEW = 'RENEW',
}

export enum AuditEntityType {
  BOOK = 'books',
  USER = 'users',
  CIRCULATION = 'circulation',
  PRESERVATION = 'preservation',
  BARCODE = 'barcode',
  CATEGORY = 'categories',
  SETTINGS = 'settings',
  REPORT = 'reports',
  SYSTEM = 'system',
}

export enum AuditStatus {
  SUCCESS = 'success',
  FAILURE = 'failure',
  WARNING = 'warning',
}

export interface AuditLog {
  id: string;
  user_id: string | null;
  username: string | null;
  user_role: string | null;
  action: AuditAction;
  entity_type: AuditEntityType;
  entity_id: string | null;
  entity_name: string | null;
  description: string;
  changes: Record<string, any> | null;
  metadata: Record<string, any>;
  ip_address: string | null;
  user_agent: string | null;
  request_method: string | null;
  request_path: string | null;
  status: AuditStatus;
  error_message: string | null;
  created_at: string;
}

export interface AuditLogFilters {
  page?: number;
  page_size?: number;
  user_id?: string;
  action?: AuditAction;
  entity_type?: AuditEntityType;
  entity_id?: string;
  status?: AuditStatus;
  start_date?: string;
  end_date?: string;
}

export interface AuditLogListResponse {
  data: AuditLog[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
}

export interface ActivitySummary {
  action: string;
  entity_type: string;
  count: number;
}

export interface AuditStatistics {
  total_activities: number;
  activities_by_action: Record<string, number>;
  activities_by_entity: Record<string, number>;
  activities_by_user: Array<{
    user_id: string;
    username: string;
    count: number;
  }>;
  recent_activities: AuditLog[];
  activity_summary: ActivitySummary[];
}

export interface UserActivity {
  date: string;
  action_count: number;
}

export interface UserActivityReport {
  user_id: string;
  username: string;
  total_actions: number;
  actions_by_type: Record<string, number>;
  daily_activity: UserActivity[];
  recent_actions: AuditLog[];
}

// Helper functions
export const getActionLabel = (action: AuditAction): string => {
  const labels: Record<AuditAction, string> = {
    [AuditAction.CREATE]: 'Created',
    [AuditAction.UPDATE]: 'Updated',
    [AuditAction.DELETE]: 'Deleted',
    [AuditAction.LOGIN]: 'Logged In',
    [AuditAction.LOGOUT]: 'Logged Out',
    [AuditAction.ACCESS]: 'Accessed',
    [AuditAction.EXPORT]: 'Exported',
    [AuditAction.IMPORT]: 'Imported',
    [AuditAction.GENERATE]: 'Generated',
    [AuditAction.SCAN]: 'Scanned',
    [AuditAction.CHECKOUT]: 'Checked Out',
    [AuditAction.CHECKIN]: 'Checked In',
    [AuditAction.RESERVE]: 'Reserved',
    [AuditAction.RENEW]: 'Renewed',
  };
  return labels[action] || action;
};

export const getActionColor = (action: AuditAction): string => {
  const colors: Record<AuditAction, string> = {
    [AuditAction.CREATE]: 'text-green-600 bg-green-50',
    [AuditAction.UPDATE]: 'text-blue-600 bg-blue-50',
    [AuditAction.DELETE]: 'text-red-600 bg-red-50',
    [AuditAction.LOGIN]: 'text-purple-600 bg-purple-50',
    [AuditAction.LOGOUT]: 'text-gray-600 bg-gray-50',
    [AuditAction.ACCESS]: 'text-cyan-600 bg-cyan-50',
    [AuditAction.EXPORT]: 'text-orange-600 bg-orange-50',
    [AuditAction.IMPORT]: 'text-yellow-600 bg-yellow-50',
    [AuditAction.GENERATE]: 'text-indigo-600 bg-indigo-50',
    [AuditAction.SCAN]: 'text-teal-600 bg-teal-50',
    [AuditAction.CHECKOUT]: 'text-pink-600 bg-pink-50',
    [AuditAction.CHECKIN]: 'text-lime-600 bg-lime-50',
    [AuditAction.RESERVE]: 'text-violet-600 bg-violet-50',
    [AuditAction.RENEW]: 'text-emerald-600 bg-emerald-50',
  };
  return colors[action] || 'text-gray-600 bg-gray-50';
};

export const getStatusColor = (status: AuditStatus): string => {
  const colors: Record<AuditStatus, string> = {
    [AuditStatus.SUCCESS]: 'text-green-600 bg-green-50',
    [AuditStatus.FAILURE]: 'text-red-600 bg-red-50',
    [AuditStatus.WARNING]: 'text-yellow-600 bg-yellow-50',
  };
  return colors[status] || 'text-gray-600 bg-gray-50';
};

File: backend/app/models/audit.py
"""
Audit Log Models
Models for activity tracking and audit trail
"""

from pydantic import BaseModel, Field
from typing import Optional, Dict, Any, List
from datetime import datetime, date
from enum import Enum
from uuid import UUID


class AuditAction(str, Enum):
    """Types of auditable actions"""
    CREATE = "CREATE"
    UPDATE = "UPDATE"
    DELETE = "DELETE"
    LOGIN = "LOGIN"
    LOGOUT = "LOGOUT"
    ACCESS = "ACCESS"
    EXPORT = "EXPORT"
    IMPORT = "IMPORT"
    GENERATE = "GENERATE"
    SCAN = "SCAN"
    CHECKOUT = "CHECKOUT"
    CHECKIN = "CHECKIN"
    RESERVE = "RESERVE"
    RENEW = "RENEW"


class AuditEntityType(str, Enum):
    """Types of entities that can be audited"""
    BOOK = "books"
    USER = "users"
    CIRCULATION = "circulation"
    PRESERVATION = "preservation"
    BARCODE = "barcode"
    CATEGORY = "categories"
    SETTINGS = "settings"
    REPORT = "reports"
    SYSTEM = "system"


class AuditStatus(str, Enum):
    """Status of the audited action"""
    SUCCESS = "success"
    FAILURE = "failure"
    WARNING = "warning"


class AuditLogBase(BaseModel):
    """Base audit log fields"""
    user_id: Optional[UUID] = None
    action: AuditAction
    entity_type: AuditEntityType
    entity_id: Optional[UUID] = None
    entity_name: Optional[str] = None
    description: str
    changes: Optional[Dict[str, Any]] = None
    metadata: Dict[str, Any] = {}
    ip_address: Optional[str] = None
    user_agent: Optional[str] = None
    request_method: Optional[str] = None
    request_path: Optional[str] = None
    status: AuditStatus = AuditStatus.SUCCESS
    error_message: Optional[str] = None


class AuditLogCreate(AuditLogBase):
    """Schema for creating audit log entry"""
    pass

    class Config:
        json_schema_extra = {
            "example": {
                "user_id": "123e4567-e89b-12d3-a456-426614174000",
                "action": "UPDATE",
                "entity_type": "books",
                "entity_id": "123e4567-e89b-12d3-a456-426614174001",
                "entity_name": "Historic Manuscript",
                "description": "Updated book metadata",
                "changes": {
                    "title": {"old": "Old Title", "new": "New Title"}
                },
                "status": "success"
            }
        }


class AuditLogResponse(AuditLogBase):
    """Schema for audit log response"""
    id: UUID
    username: Optional[str] = None
    user_role: Optional[str] = None
    created_at: datetime

    class Config:
        from_attributes = True
        json_schema_extra = {
            "example": {
                "id": "123e4567-e89b-12d3-a456-426614174002",
                "user_id": "123e4567-e89b-12d3-a456-426614174000",
                "username": "admin@ministry.om",
                "user_role": "administrator",
                "action": "UPDATE",
                "entity_type": "books",
                "entity_id": "123e4567-e89b-12d3-a456-426614174001",
                "entity_name": "Historic Manuscript",
                "description": "Updated book metadata",
                "status": "success",
                "created_at": "2024-11-14T10:00:00Z"
            }
        }


class AuditLogFilters(BaseModel):
    """Filters for querying audit logs"""
    user_id: Optional[UUID] = None
    action: Optional[AuditAction] = None
    entity_type: Optional[AuditEntityType] = None
    entity_id: Optional[UUID] = None
    status: Optional[AuditStatus] = None
    start_date: Optional[datetime] = None
    end_date: Optional[datetime] = None
    page: int = Field(default=1, ge=1)
    page_size: int = Field(default=20, ge=1, le=100)


class AuditLogListResponse(BaseModel):
    """Paginated list of audit logs"""
    data: List[AuditLogResponse]
    total: int
    page: int
    page_size: int
    total_pages: int


class ActivitySummary(BaseModel):
    """Summary of activities by type"""
    action: str
    entity_type: str
    count: int


class AuditStatistics(BaseModel):
    """Audit log statistics"""
    total_activities: int
    activities_by_action: Dict[str, int]
    activities_by_entity: Dict[str, int]
    activities_by_user: List[Dict[str, Any]]  # Top users by activity
    recent_activities: List[AuditLogResponse]
    activity_summary: List[ActivitySummary]

    class Config:
        json_schema_extra = {
            "example": {
                "total_activities": 1500,
                "activities_by_action": {
                    "CREATE": 500,
                    "UPDATE": 600,
                    "DELETE": 50,
                    "LOGIN": 350
                },
                "activities_by_entity": {
                    "books": 800,
                    "users": 400,
                    "circulation": 300
                },
                "activities_by_user": [
                    {"user_id": "...", "username": "admin@ministry.om", "count": 250},
                    {"user_id": "...", "username": "librarian@ministry.om", "count": 180}
                ],
                "recent_activities": [],
                "activity_summary": []
            }
        }


class UserActivity(BaseModel):
    """User activity over time"""
    date: date
    action_count: int


class UserActivityReport(BaseModel):
    """Detailed user activity report"""
    user_id: UUID
    username: str
    total_actions: int
    actions_by_type: Dict[str, int]
    daily_activity: List[UserActivity]
    recent_actions: List[AuditLogResponse]
